### PR TITLE
simplify api registration

### PR DESCRIPTION
**Pull request by deads2k**
The current registration and groupmeta is only use to determine a preferred ordering to versions.  The scheme already knows about all versions, so this simply makes that knowledge official.  After doing that, the announce, registered, and apimachinery/pkg/apimachinery all drop out.

With this change we'll be able to create `install` packages for each external apigroup that accept a scheme and have suggested orderings.  This will make it possible to close the consistency gap with kubectl, client, and apiserver.

@kubernetes/sig-api-machinery-pr-reviews 
@lavalamp @smarterclayton @liggitt @sttts 

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubectl/cmd/apply_set_last_applied.go [hunk 1]
```diff
@@ -151,7 +151,7 @@ func (o *SetLastAppliedOptions) Validate() error {
 		if err != nil {
 			return err
 		}
-		patchBuf, diffBuf, patchType, err := editor.GetApplyPatch(info.Object, scheme.DefaultJSONEncoder())
+		patchBuf, diffBuf, patchType, err := editor.GetApplyPatch(info.Object.(runtime.Unstructured))
```
Comment (yliaog) [inline pkg/kubectl/cmd/apply_set_last_applied.go]: this commit does not belong to this PR, does it?

Comment (deads2k) [reply pkg/kubectl/cmd/apply_set_last_applied.go]: > this commit does not belong to this PR, does it?

Split out and merged here: https://github.com/kubernetes/kubernetes/pull/63398 .  Sorry about the noise.

File: pkg/client/clientset_generated/internalclientset/typed/apps/internalversion/apps_client.go [hunk 2]
```diff
@@ -71,17 +71,12 @@ func New(c rest.Interface) *AppsClient {
 }
 
 func setConfigDefaults(config *rest.Config) error {
```
Comment (liggitt) [inline pkg/client/clientset_generated/internalclientset/typed/apps/internalversion/apps_client.go]: these changes weren't generated?

File: pkg/kubeapiserver/options/storage_versions.go [hunk 3]
```diff
@@ -104,3 +106,14 @@ func (s *StorageSerializationOptions) AddFlags(fs *pflag.FlagSet) {
 		"It defaults to a list of preferred versions of all known groups.")
 
 }
+
+// AllPreferredGroupVersions returns the preferred versions of all registered
+// groups in the form of "group1/version1,group2/version2,..."
+func ToPreferredVersionString(versions []schema.GroupVersion) string {
```
Comment (liggitt) [inline pkg/kubeapiserver/options/storage_versions.go]: is order not important here? or is it an error for versions to contain more than one item in the same group?

Comment (deads2k) [reply pkg/kubeapiserver/options/storage_versions.go]: > is order not important here? or is it an error for versions to contain more than one item in the same group?

Groups only have one version when calling it.

File: pkg/client/clientset_generated/internalclientset/typed/apps/internalversion/apps_client.go (hunk 2 above)
Comment (deads2k) [reply pkg/client/clientset_generated/internalclientset/typed/apps/internalversion/apps_client.go]: > these changes weren't generated?

They were (otherwise verify fails).

File: pkg/kubectl/scheme/install.go [hunk 4]
```diff
@@ -55,238 +55,24 @@ import (
...
-		},
-		announced.VersionToSchemeFunc{
-			storagev1.SchemeGroupVersion.Version:      storagev1.AddToScheme,
-			storagev1beta1.SchemeGroupVersion.Version: storagev1beta1.AddToScheme,
-		},
-	).Register(Registry, Scheme); err != nil {
-		panic(err)
-	}
+	Scheme.SetVersionPriorityOrDie(corev1.SchemeGroupVersion)
+	Scheme.SetVersionPriorityOrDie(admissionv1alpha1.SchemeGroupVersion)
```
Comment (liggitt) [inline pkg/kubectl/scheme/install.go]: I was surprised these weren't done in `k8s.io/client-go/kubernetes/scheme#AddToScheme()`, at least as an initial ordering

**Comment by liggitt** [comment]
nit: pick prioritized or preferred (PrioritizedVersionsForGroup vs PreferredVersionsAllGroups)

File: staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go [hunk 5]
```diff
@@ -620,3 +628,69 @@ func setTargetKind(obj Object, kind schema.GroupVersionKind) {
 	}
 	obj.GetObjectKind().SetGroupVersionKind(kind)
 }
+
+// this only fails on programmer error
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go]: doc all specified versions must be in the same group, and the specified order overwrites any previously specified order for this group

File: pkg/kubectl/scheme/install.go (hunk 4 above)
Comment (deads2k) [reply pkg/kubectl/scheme/install.go]: > I was surprised all of these weren't done in k8s.io/client-go/kubernetes/scheme#AddToScheme(), at least as an initial ordering

That one is generated, isn't it?  Also, the generated client knows what its expects, so preference there doesn't really mean anything.  The real question is why anyone would try to wire up to that client....

File: pkg/client/clientset_generated/internalclientset/typed/apps/internalversion/apps_client.go (hunk 2 above)
Comment (liggitt) [reply pkg/client/clientset_generated/internalclientset/typed/apps/internalversion/apps_client.go]: ok, saw a "generated" commit split out and thought this commit was limited to manual changes

**Comment by deads2k** [comment]
rebased

File: staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go [hunk 6]
```diff
@@ -620,3 +628,69 @@ func setTargetKind(obj Object, kind schema.GroupVersionKind) {
...
+	for group, versions := range s.versionPriority {
+		for _, version := range versions {
+			ret = append(ret, schema.GroupVersion{Group: group, Version: version})
+			break
+		}
+	}
+	return ret
+}
+
+func (s *Scheme) IsGroupRegistered(group string) bool {
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go]: the following seems very strange to me:

```go
gv := schema.GroupVersion{Group:"foo",Version:"v1"}
scheme.AddKnownTypes(gv, ...)
scheme.KnownTypes(gv) // returns results
scheme.IsGroupRegistered(gv.Group) // returns false
```

if you register types for a few versions of a group and do not set an explicit priority order, then ask for the priority order, I'd expect either to return the first seen order of the versions, or error. I wouldn't expect the returned list to not contain the group/versions the scheme knew about

likewise, if I register KnownTypes for 4 versions of a group, then set an explicit priority of 2 of those versions, I'd either expect the other two versions to go at the end in indeterminate order, or to error. I wouldn't expect the other 2 versions to disappear from the ordered list of all versions.

File: pkg/kubectl/scheme/install.go (hunk 4 above)
Comment (liggitt) [reply pkg/kubectl/scheme/install.go]: hmm... is the setting of priority order here just to make `IsGroupRegistered` and `IsVersionRegistered` do reasonable things? (c.f. https://github.com/kubernetes/kubernetes/pull/63364/files#r186214718)

**Comment by deads2k** [comment]
> nit: pick prioritized or preferred (PrioritizedVersionsForGroup vs PreferredVersionsAllGroups)

The two are different.  Preferred only picks the ones with highest priority.  Will doc

**Comment by deads2k** [comment]
comments addressed.

File: pkg/client/clientset_generated/internalclientset/typed/apps/internalversion/apps_client.go (hunk 2 above)
Comment (deads2k) [reply pkg/client/clientset_generated/internalclientset/typed/apps/internalversion/apps_client.go]: > ok, saw a "generated" commit split out and thought this commit was limited to manual changes

fixed

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/register.go [hunk 7]
```diff
@@ -37,8 +37,9 @@ func Resource(resource string) schema.GroupResource {
 }
 
 var (
-	SchemeBuilder = runtime.NewSchemeBuilder(addKnownTypes)
-	AddToScheme   = SchemeBuilder.AddToScheme
+	SchemeBuilder    = runtime.NewSchemeBuilder(addKnownTypes)
+	AddToScheme      = SchemeBuilder.AddToScheme
+	AddToSchemeOrDie = SchemeBuilder.AddToSchemeOrDie
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/register.go]: would prefer some compositional or even local "or die" logic instead of defining it in every group.

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/register.go]: This can be a follow-up, with all the good this PR does.

Comment (deads2k) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/register.go]: Addressing thsi now

**Comment by deads2k** [comment]
now with `Must`

File: pkg/printers/customcolumn_flags.go [hunk 8]
```diff
@@ -84,7 +84,7 @@ func (f *CustomColumnsPrintFlags) ToPrinter(templateFormat string) (ResourcePrin
 	}
 
 	p, err := NewCustomColumnsPrinterFromSpec(templateValue, decoder, f.NoHeaders)
-	return NewVersionedPrinter(p, legacyscheme.Scheme, legacyscheme.Scheme, scheme.Versions...), err
+	return NewVersionedPrinter(p, legacyscheme.Scheme, legacyscheme.Scheme, scheme.Scheme.PreferredVersionsAllGroups()...), err
```
Comment (liggitt) [inline pkg/printers/customcolumn_flags.go]: all `Versions...` should be replaced with `PrioritizedVersionsAllGroups()...` to remain equivalent, right?

File: staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go [hunk 9]
```diff
@@ -620,3 +634,123 @@ func setTargetKind(obj Object, kind schema.GroupVersionKind) {
...
+	}
+	if len(groups) != 1 {
+		return fmt.Errorf("must register versions for exactly one group: %v", strings.Join(groups.List(), ", "))
+	}
+
+	s.versionPriority[groups.List()[0]] = order
+	return nil
+}
+
+func (s *Scheme) PrioritizedVersionsForGroup(group string) []schema.GroupVersion {
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go]: needs godoc (and PrioritizedVersionsAllGroups)

File: staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go [hunk 10]
```diff
@@ -620,3 +634,123 @@ func setTargetKind(obj Object, kind schema.GroupVersionKind) {
...
+			}
+		}
+		if !found {
+			ret = append(ret, observedVersion)
+		}
+	}
+	return ret
+}
+
+// PreferredVersionsAllGroups returns ok the most preferred version for every group.
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go]: typo in doc

File: staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go [hunk 11]
```diff
@@ -620,3 +634,123 @@ func setTargetKind(obj Object, kind schema.GroupVersionKind) {
...
+
+//  all specified versions must be in the same group, and the specified order overwrites any previously specified order for this group
+func (s *Scheme) SetVersionPriority(versions ...schema.GroupVersion) error {
+	groups := sets.String{}
+	order := []string{}
+	for _, version := range versions {
+		groups.Insert(version.Group)
+		order = append(order, version.Version)
+	}
+	if len(groups) != 1 {
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go]: would also be useful to ensure an empty or internal version wasn't passed to this function

File: staging/src/k8s.io/apiserver/pkg/server/resourceconfig/helpers.go [hunk 12]
```diff
@@ -30,11 +30,11 @@ import (
 // GroupVersionRegistry provides access to registered group versions.
 type GroupVersionRegistry interface {
 	// IsRegistered returns true if given group is registered.
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/resourceconfig/helpers.go]: update docs to match renamed functions

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_group.go [hunk 13]
```diff
@@ -210,17 +210,12 @@ func New(c $.restRESTClientInterface|raw$) *$.GroupGoName$$.Version$Client {
...
-		return err
-	}
-
 	config.APIPath = $.apiPath$
 	if config.UserAgent == "" {
 		config.UserAgent = $.restDefaultKubernetesUserAgent|raw$()
 	}
-	if config.GroupVersion == nil || config.GroupVersion.Group != g.GroupVersions[0].Group {
-		gv := g.GroupVersions[0]
+	if config.GroupVersion == nil || config.GroupVersion.Group != scheme.Scheme.PrioritizedVersionsForGroup("$.groupName$")[0].Group {
```
Comment (liggitt) [inline staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_group.go]: worth checking if len is 0 and erroring rather than panicing?

**Comment by liggitt** [comment]
the `scheme.Versions...` transformation is the only substantive comment, the rest are nits. a few generated files bled into the non-generated commit if you care.

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_group.go (hunk 13 above)
Comment (deads2k) [reply staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_group.go]: > worth checking if len is 0 and erroring rather than panicing?

If we got the scheme wrong somehow?  I don't think so.  That's a meta developer error.

Comment (liggitt) [reply staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_group.go]: ah, missed that the referenced scheme was generated of a piece. that's fine, then

**Comment by deads2k** [comment]
stackdriver seems to having some trouble today.

/retest

**Comment by lavalamp** [comment]
looks like a great change!

**Comment by lavalamp** [comment]
/hold

for nits.

/approve

**Comment by deads2k** [comment]
comments addressed.  Minor rebase.  retagging

/hold cancel
